File: tests/source/go/packages/sdk/go/envelope_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Unit tests for wire envelope encoding, decoding, and baggage parsing.

package sdk_test

import (
	"net/http"
	"strings"
	"testing"

	sdk "github.com/garudex-labs/caracal/packages/sdk/go"
)

func TestParseTraceparentValid(t *testing.T) {
	traceID, flags := sdk.ParseTraceparent("00-0123456789abcdef0123456789abcdef-aabbccddeeff0011-01")
	if traceID != "0123456789abcdef0123456789abcdef" {
		t.Fatalf("unexpected trace id: %q", traceID)
	}
	if flags != "01" {
		t.Fatalf("unexpected flags: %q", flags)
	}
}

func TestParseTraceparentFutureVersion(t *testing.T) {
	traceID, flags := sdk.ParseTraceparent("01-0123456789abcdef0123456789abcdef-aabbccddeeff0011-00-extrafield")
	if traceID != "0123456789abcdef0123456789abcdef" {
		t.Fatalf("future version with extra fields must parse, got %q", traceID)
	}
	if flags != "00" {
		t.Fatalf("unexpected flags: %q", flags)
	}
}

func TestParseTraceparentEmpty(t *testing.T) {
	if traceID, _ := sdk.ParseTraceparent(""); traceID != "" {
		t.Error("empty input must return empty")
	}
}

func TestParseTraceparentAllZero(t *testing.T) {
	if traceID, _ := sdk.ParseTraceparent("00-00000000000000000000000000000000-aabbccddeeff0011-01"); traceID != "" {
		t.Error("all-zero trace id must return empty")
	}
	if traceID, _ := sdk.ParseTraceparent("00-0123456789abcdef0123456789abcdef-0000000000000000-01"); traceID != "" {
		t.Error("all-zero span id must return empty")
	}
}

func TestParseTraceparentInvalidFormats(t *testing.T) {
	cases := []string{
		"invalid",
		"00-short-aabbccddeeff0011-01",
		"00-0123456789abcdef0123456789ABCDEF-aabbccddeeff0011-01", // uppercase
		"ff-0123456789abcdef0123456789abcdef-aabbccddeeff0011-01", // forbidden version
		"00-0123456789abcdef0123456789abcdef-aabbccddeeff0011-01-extra", // version 00 with extra field
		"noteven",
	}
	for _, v := range cases {
		if traceID, _ := sdk.ParseTraceparent(v); traceID != "" {
			t.Errorf("expected empty for %q", v)
		}
	}
}

func TestParseBaggageKeyValue(t *testing.T) {
	bag := sdk.ParseBaggage("k1=v1,k2=v2")
	if bag["k1"] != "v1" || bag["k2"] != "v2" {
		t.Errorf("unexpected baggage: %v", bag)
	}
}

func TestParseBaggagePercentEncoded(t *testing.T) {
	bag := sdk.ParseBaggage("k=hello%20world")
	if bag["k"] != "hello world" {
		t.Errorf("expected decoded value, got %q", bag["k"])
	}
}

func TestParseBaggagePlusIsLiteral(t *testing.T) {
	bag := sdk.ParseBaggage("k=a+b")
	if bag["k"] != "a+b" {
		t.Errorf("'+' must stay literal per W3C Baggage, got %q", bag["k"])
	}
}

func TestParseBaggageMalformedPercentKeptRaw(t *testing.T) {
	bag := sdk.ParseBaggage("k=broken%zz")
	if bag["k"] != "broken%zz" {
		t.Errorf("malformed percent escape must stay raw, got %q", bag["k"])
	}
}

func TestParseBaggageMalformedSkipped(t *testing.T) {
	bag := sdk.ParseBaggage("noequal,k=v")
	if _, ok := bag["noequal"]; ok {
		t.Error("malformed entry without '=' must be skipped")
	}
	if bag["k"] != "v" {
		t.Errorf("valid entry must still parse, got %q", bag["k"])
	}
}

func TestParseBaggageEmpty(t *testing.T) {
	if len(sdk.ParseBaggage("")) != 0 {
		t.Error("empty string must produce empty map")
	}
}

func TestParseBaggageOversizeDiscarded(t *testing.T) {
	if len(sdk.ParseBaggage("k=" + strings.Repeat("a", 9000))) != 0 {
		t.Error("headers above the W3C size limit must be discarded")
	}
	members := make([]string, 65)
	for i := range members {
		members[i] = "k" + string(rune('a'+i%26)) + "=v"
	}
	if len(sdk.ParseBaggage(strings.Join(members, ","))) != 0 {
		t.Error("headers above the W3C member limit must be discarded")
	}
}

func TestParseBaggageSemicolonStripsProperties(t *testing.T) {
	bag := sdk.ParseBaggage("k=v;property=ignored")
	if bag["k"] != "v" {
		t.Errorf("semicolon properties must be stripped, got %q", bag["k"])
	}
}

func TestEncodeBaggageDeterministicOrder(t *testing.T) {
	got := sdk.EncodeBaggage(map[string]string{"zeta": "1", "alpha": "2", "mid": "3"})
	if got != "alpha=2,mid=3,zeta=1" {
		t.Errorf("baggage must encode in sorted key order, got %q", got)
	}
}

func TestEncodeBaggagePercentEncodesReserved(t *testing.T) {
	got := sdk.EncodeBaggage(map[string]string{"k": "a b%c,d=e"})
	if got != "k=a%20b%25c%2Cd%3De" {
		t.Errorf("reserved characters must percent-encode, got %q", got)
	}
	back := sdk.ParseBaggage(got)
	if back["k"] != "a b%c,d=e" {
		t.Errorf("round trip mismatch: %q", back["k"])
	}
}

func TestEncodeEnvelopeNeverEmitsAuthorization(t *testing.T) {
	env := sdk.Envelope{SubjectToken: "tok", TraceID: "0123456789abcdef0123456789abcdef", Hop: 1}
	got := map[string]string{}
	sdk.EncodeEnvelope(env, func(k, v string) { got[k] = v }, nil)
	if _, ok := got[sdk.HeaderAuthorization]; ok {
		t.Error("encode must never emit Authorization; credential placement is a client-layer decision")
	}
}

func TestDecodeEncodeEnvelopeRoundTrip(t *testing.T) {
	env := sdk.Envelope{
		AgentSessionID:   "sess1",
		DelegationEdgeID: "edge1",
		ParentEdgeID:     "parent1",
		SessionID:        "sid1",
		TraceID:          "0123456789abcdef0123456789abcdef",
		TraceFlags:       "00",
		TraceState:       "vendor=value",
		Baggage:          map[string]string{"tenant": "pied-piper"},
		Hop:              3,
	}
	headers := map[string]string{}
	sdk.EncodeEnvelope(env, func(k, v string) { headers[k] = v }, nil)

	out := sdk.DecodeEnvelope(func(k string) string { return headers[k] })

	if out.AgentSessionID != env.AgentSessionID {
		t.Errorf("AgentSessionID mismatch: %q vs %q", out.AgentSessionID, env.AgentSessionID)
	}
	if out.DelegationEdgeID != env.DelegationEdgeID {
		t.Errorf("DelegationEdgeID mismatch")
	}
	if out.ParentEdgeID != env.ParentEdgeID {
		t.Errorf("ParentEdgeID mismatch")
	}
	if out.SessionID != env.SessionID {
		t.Errorf("SessionID mismatch")
	}
	if out.TraceID != env.TraceID {
		t.Errorf("TraceID mismatch: %q vs %q", out.TraceID, env.TraceID)
	}
	if out.TraceFlags != env.TraceFlags {
		t.Errorf("TraceFlags mismatch: %q vs %q", out.TraceFlags, env.TraceFlags)
	}
	if out.TraceState != env.TraceState {
		t.Errorf("TraceState mismatch: %q vs %q", out.TraceState, env.TraceState)
	}
	if out.Baggage["tenant"] != "pied-piper" {
		t.Errorf("third-party baggage must round trip, got %v", out.Baggage)
	}
	if out.Hop != env.Hop {
		t.Errorf("Hop mismatch: %d vs %d", out.Hop, env.Hop)
	}
}

func TestDecodeEnvelopeBearerPrefix(t *testing.T) {
	cases := map[string]string{
		"Bearer mytoken":      "mytoken",
		"bearer mytoken":      "mytoken",
		"BEARER   mytoken   ": "mytoken",
	}
	for raw, want := range cases {
		env := sdk.DecodeEnvelope(func(k string) string {
			if k == sdk.HeaderAuthorization {
				return raw
			}
			return ""
		})
		if env.SubjectToken != want {
			t.Errorf("%q: expected %q, got %q", raw, want, env.SubjectToken)
		}
	}
}

func TestDecodeEnvelopeNonBearerIgnored(t *testing.T) {
	for _, raw := range []string{"Basic dXNlcjpwYXNz", "Bearer", "Bearer ", "Bearertok"} {
		env := sdk.DecodeEnvelope(func(k string) string {
			if k == sdk.HeaderAuthorization {
				return raw
			}
			return ""
		})
		if env.SubjectToken != "" {
			t.Errorf("%q must be ignored, got %q", raw, env.SubjectToken)
		}
	}
}

func TestHopClamping(t *testing.T) {
	cases := []struct {
		raw  string
		want int
	}{
		{"0", 0},
		{"1", 1},
		{"10", sdk.MaxHop},
		{"11", sdk.MaxHop},
		{"100", sdk.MaxHop},
		{"99999999999999999999", sdk.MaxHop},
		{"-1", 0},
		{"-99", 0},
		{"+3", 0},
		{"3x", 0},
		{"1e2", 0},
	}
	for _, tc := range cases {
		env := sdk.DecodeEnvelope(func(k string) string {
			if k == sdk.HeaderBaggage {
				return sdk.BaggageHop + "=" + tc.raw
			}
			return ""
		})
		if env.Hop != tc.want {
			t.Errorf("hop=%q: want %d got %d", tc.raw, tc.want, env.Hop)
		}
	}
}

func TestEncodeEnvelopeOmitsHopZeroWithoutIdentityFields(t *testing.T) {
	got := map[string]string{}
	sdk.EncodeEnvelope(sdk.Envelope{TraceID: "0123456789abcdef0123456789abcdef"}, func(k, v string) { got[k] = v }, nil)
	if _, ok := got[sdk.HeaderBaggage]; ok {
		t.Errorf("baggage must be omitted for a root envelope, got %q", got[sdk.HeaderBaggage])
	}
}

func TestEncodeEnvelopeMergePreservesExistingHeaders(t *testing.T) {
	existing := map[string]string{
		sdk.HeaderTraceparent: "00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab-bbbbbbbbbbbbbbbb-01",
		sdk.HeaderTracestate:  "otel=span",
		sdk.HeaderBaggage:     "tenant=hooli," + sdk.BaggageHop + "=9",
	}
	env := sdk.Envelope{
		AgentSessionID: "sess",
		TraceID:        "0123456789abcdef0123456789abcdef",
		TraceState:     "caracal=ignored",
		Hop:            2,
	}
	sdk.EncodeEnvelope(env, func(k, v string) { existing[k] = v }, func(k string) string { return existing[k] })

	if existing[sdk.HeaderTraceparent] != "00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab-bbbbbbbbbbbbbbbb-01" {
		t.Errorf("an existing valid traceparent must win, got %q", existing[sdk.HeaderTraceparent])
	}
	if existing[sdk.HeaderTracestate] != "otel=span" {
		t.Errorf("an existing tracestate must win, got %q", existing[sdk.HeaderTracestate])
	}
	bag := sdk.ParseBaggage(existing[sdk.HeaderBaggage])
	if bag["tenant"] != "hooli" {
		t.Errorf("existing third-party baggage must survive, got %v", bag)
	}
	if bag[sdk.BaggageAgentSession] != "sess" || bag[sdk.BaggageHop] != "2" {
		t.Errorf("envelope caracal.* fields must win over stale entries, got %v", bag)
	}
}

func TestInjectFromHTTPRequestRoundTrip(t *testing.T) {
	env := sdk.Envelope{
		AgentSessionID: "sess",
		TraceID:        "abcdef0123456789abcdef0123456789",
		Hop:            2,
	}
	h := http.Header{}
	sdk.InjectHTTP(env, h)

	req, _ := http.NewRequest("GET", "http://example.com", nil)
	req.Header = h
	out := sdk.FromHTTPRequest(req)

	if out.AgentSessionID != env.AgentSessionID {
		t.Errorf("AgentSessionID: %q vs %q", out.AgentSessionID, env.AgentSessionID)
	}
	if out.TraceID != env.TraceID {
		t.Errorf("TraceID: %q vs %q", out.TraceID, env.TraceID)
	}
	if out.Hop != env.Hop {
		t.Errorf("Hop: %d vs %d", out.Hop, env.Hop)
	}
}

func TestFromHTTPRequestJoinsRepeatedBaggageHeaders(t *testing.T) {
	req, _ := http.NewRequest("GET", "http://example.com", nil)
	req.Header.Add("Baggage", sdk.BaggageAgentSession+"=sess")
	req.Header.Add("Baggage", "tenant=hooli")
	out := sdk.FromHTTPRequest(req)
	if out.AgentSessionID != "sess" {
		t.Errorf("first baggage header lost: %q", out.AgentSessionID)
	}
	if out.Baggage["tenant"] != "hooli" {
		t.Errorf("second baggage header lost: %v", out.Baggage)
	}
}

func TestToMapFromMapRoundTrip(t *testing.T) {
	env := sdk.Envelope{
		AgentSessionID:   "sess",
		DelegationEdgeID: "edge",
		SessionID:        "sid",
		TraceID:          "0123456789abcdef0123456789abcdef",
		Hop:              1,
	}
	m := sdk.ToHeaders(env)
	out := sdk.FromHeaders(m)

	if out.AgentSessionID != env.AgentSessionID {
		t.Errorf("AgentSessionID mismatch")
	}
	if out.DelegationEdgeID != env.DelegationEdgeID {
		t.Errorf("DelegationEdgeID mismatch")
	}
	if out.SessionID != env.SessionID {
		t.Errorf("SessionID mismatch")
	}
	if out.TraceID != env.TraceID {
		t.Errorf("TraceID mismatch")
	}
}

func TestFromHeadersCaseInsensitive(t *testing.T) {
	out := sdk.FromHeaders(map[string]string{"AUTHORIZATION": "Bearer tok"})
	if out.SubjectToken != "tok" {
		t.Errorf("case-insensitive key lookup failed: %q", out.SubjectToken)
	}
}

func TestEncodeEnvelopeGeneratesTraceIDWhenMissing(t *testing.T) {
	env := sdk.Envelope{Hop: 1}
	got := map[string]string{}
	sdk.EncodeEnvelope(env, func(k, v string) { got[k] = v }, nil)

	tp := got[sdk.HeaderTraceparent]
	if traceID, _ := sdk.ParseTraceparent(tp); traceID == "" {
		t.Errorf("encode must generate a traceparent when TraceID is empty, got %q", tp)
	}
}

func TestEncodeEnvelopePropagatesTraceFlags(t *testing.T) {
	env := sdk.Envelope{TraceID: "0123456789abcdef0123456789abcdef", TraceFlags: "00", Hop: 1}
	got := map[string]string{}
	sdk.EncodeEnvelope(env, func(k, v string) { got[k] = v }, nil)
	if _, flags := sdk.ParseTraceparent(got[sdk.HeaderTraceparent]); flags != "00" {
		t.Errorf("trace flags must propagate, got %q", got[sdk.HeaderTraceparent])
	}
}

File: tests/typescript/unit/sdk/envelope.test.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * Unit tests for W3C Trace Context and Baggage envelope encode/decode functions.
 */

import { describe, it, expect } from 'vitest'
import {
  BaggageAgentSession,
  BaggageDelegationEdge,
  BaggageHop,
  HeaderAuthorization,
  HeaderBaggage,
  HeaderTraceparent,
  HeaderTracestate,
  MaxHop,
  decodeEnvelope,
  encodeBaggage,
  encodeEnvelope,
  fromHeaders,
  parseBaggage,
  parseTraceparent,
  toHeaders,
  type Envelope,
} from '../../../../packages/sdk/ts/src/envelope.js'

describe('parseTraceparent', () => {
  it('returns trace id and flags from a valid header', () => {
    expect(parseTraceparent('00-0123456789abcdef0123456789abcdef-0011223344556677-01')).toEqual({
      traceId: '0123456789abcdef0123456789abcdef',
      flags: '01',
    })
  })

  it('accepts future versions with extra fields', () => {
    expect(parseTraceparent('01-0123456789abcdef0123456789abcdef-0011223344556677-00-extra')).toEqual({
      traceId: '0123456789abcdef0123456789abcdef',
      flags: '00',
    })
  })

  it('rejects invalid formats', () => {
    for (const value of [
      '',
      'not-a-traceparent',
      'ff-0123456789abcdef0123456789abcdef-0011223344556677-01',
      '00-0123456789abcdef0123456789abcdef-0011223344556677-01-extra',
      '00-' + '0'.repeat(32) + '-0011223344556677-01',
      '00-0123456789abcdef0123456789abcdef-0000000000000000-01',
      '00-0123456789ABCDEF0123456789abcdef-0011223344556677-01',
    ]) {
      expect(parseTraceparent(value)).toBeUndefined()
    }
  })
})

describe('baggage encoding', () => {
  it('encodes keys in sorted order', () => {
    expect(encodeBaggage({ zeta: '1', alpha: '2', mid: '3' })).toBe('alpha=2,mid=3,zeta=1')
  })

  it('round-trips percent-encoded reserved characters', () => {
    const encoded = encodeBaggage({ k: 'a b%c,d=e' })
    expect(encoded).toBe('k=a%20b%25c%2Cd%3De')
    expect(parseBaggage(encoded)).toEqual({ k: 'a b%c,d=e' })
  })

  it('keeps + literal per W3C Baggage', () => {
    expect(parseBaggage('k=a+b')).toEqual({ k: 'a+b' })
  })

  it('discards headers above the W3C size limits', () => {
    expect(parseBaggage('k=' + 'a'.repeat(9000))).toEqual({})
    const oversized = Array.from({ length: 65 }, (_, i) => `k${i}=v`).join(',')
    expect(parseBaggage(oversized)).toEqual({})
  })
})

describe('decodeEnvelope', () => {
  it('accepts bearer tokens case-insensitively and trimmed', () => {
    for (const raw of ['Bearer tok', 'bearer tok', ' BEARER   tok  ']) {
      const env = decodeEnvelope((n) => (n === HeaderAuthorization ? raw : undefined))
      expect(env.subjectToken).toBe('tok')
    }
  })

  it('ignores non-bearer authorization values', () => {
    for (const raw of ['Basic dXNlcjpwYXNz', 'Bearer', 'Bearer ', 'Bearertok']) {
      const env = decodeEnvelope((n) => (n === HeaderAuthorization ? raw : undefined))
      expect(env.subjectToken).toBeUndefined()
    }
  })

  it('clamps hop and rejects non-digit values', () => {
    const cases: [string, number][] = [
      ['0', 0],
      ['1', 1],
      ['10', MaxHop],
      ['11', MaxHop],
      ['99999999999999999999', MaxHop],
      ['-1', 0],
      ['+3', 0],
      ['3x', 0],
      ['1e2', 0],
      ['1.5', 0],
    ]
    for (const [raw, want] of cases) {
      const env = decodeEnvelope((n) => (n === HeaderBaggage ? `${BaggageHop}=${raw}` : undefined))
      expect(env.hop, `hop=${raw}`).toBe(want)
    }
  })

  it('captures third-party baggage, tracestate, and trace flags', () => {
    const headers: Record<string, string> = {
      [HeaderTraceparent]: '00-0123456789abcdef0123456789abcdef-0011223344556677-00',
      [HeaderTracestate]: 'vendor=value',
      [HeaderBaggage]: `tenant=hooli,${BaggageHop}=1`,
    }
    const env = decodeEnvelope((n) => headers[n])
    expect(env.traceFlags).toBe('00')
    expect(env.traceState).toBe('vendor=value')
    expect(env.baggage).toEqual({ tenant: 'hooli' })
    expect(env.hop).toBe(1)
  })
})

describe('encodeEnvelope', () => {
  it('never emits Authorization', () => {
    const env: Envelope = { subjectToken: 'tok', traceId: '0123456789abcdef0123456789abcdef', hop: 1 }
    const out: Record<string, string> = {}
    encodeEnvelope(env, (k, v) => {
      out[k] = v
    })
    expect(out[HeaderAuthorization]).toBeUndefined()
  })

  it('omits baggage for a root envelope', () => {
    const out: Record<string, string> = {}
    encodeEnvelope({ traceId: '0123456789abcdef0123456789abcdef' }, (k, v) => {
      out[k] = v
    })
    expect(out[HeaderBaggage]).toBeUndefined()
    expect(out[HeaderTraceparent]).toBeDefined()
  })

  it('merges with existing headers instead of clobbering them', () => {
    const existing: Record<string, string> = {
      [HeaderTraceparent]: '00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab-bbbbbbbbbbbbbbbb-01',
      [HeaderTracestate]: 'otel=span',
      [HeaderBaggage]: `tenant=hooli,${BaggageHop}=9`,
    }
    const env: Envelope = {
      agentSessionId: 'sess',
      traceId: '0123456789abcdef0123456789abcdef',
      traceState: 'caracal=ignored',
      hop: 2,
    }
    encodeEnvelope(
      env,
      (k, v) => {
        existing[k] = v
      },
      (k) => existing[k],
    )
    expect(existing[HeaderTraceparent]).toBe('00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab-bbbbbbbbbbbbbbbb-01')
    expect(existing[HeaderTracestate]).toBe('otel=span')
    const bag = parseBaggage(existing[HeaderBaggage])
    expect(bag.tenant).toBe('hooli')
    expect(bag[BaggageAgentSession]).toBe('sess')
    expect(bag[BaggageHop]).toBe('2')
    expect(bag[BaggageDelegationEdge]).toBeUndefined()
  })

  it('propagates trace flags into the generated traceparent', () => {
    const out: Record<string, string> = {}
    encodeEnvelope({ traceId: '0123456789abcdef0123456789abcdef', traceFlags: '00', hop: 1 }, (k, v) => {
      out[k] = v
    })
    expect(parseTraceparent(out[HeaderTraceparent])?.flags).toBe('00')
  })
})

describe('toHeaders/fromHeaders', () => {
  it('round-trips the full envelope without the subject token', () => {
    const env: Envelope = {
      subjectToken: 'tok',
      agentSessionId: 'agent-1',
      delegationEdgeId: 'edge-1',
      parentEdgeId: 'parent-1',
      sessionId: 'sid-1',
      traceId: 'a'.repeat(32),
      traceFlags: '00',
      traceState: 'vendor=value',
      baggage: { tenant: 'hooli' },
      hop: 2,
    }
    const headers = toHeaders(env)
    expect(headers[HeaderAuthorization]).toBeUndefined()

    const recovered = fromHeaders(headers)
    expect(recovered.subjectToken).toBeUndefined()
    expect(recovered.agentSessionId).toBe('agent-1')
    expect(recovered.delegationEdgeId).toBe('edge-1')
    expect(recovered.parentEdgeId).toBe('parent-1')
    expect(recovered.sessionId).toBe('sid-1')
    expect(recovered.traceId).toBe('a'.repeat(32))
    expect(recovered.traceFlags).toBe('00')
    expect(recovered.traceState).toBe('vendor=value')
    expect(recovered.baggage).toEqual({ tenant: 'hooli' })
    expect(recovered.hop).toBe(2)
  })

  it('joins repeated baggage values case-insensitively', () => {
    const env = fromHeaders({
      AUTHORIZATION: 'Bearer tok',
      Baggage: [`${BaggageAgentSession}=sess`, 'tenant=hooli'],
    })
    expect(env.subjectToken).toBe('tok')
    expect(env.agentSessionId).toBe('sess')
    expect(env.baggage).toEqual({ tenant: 'hooli' })
  })
})
